File: AvaloniaMedia/FFmpeg/Model/Frame.cs

using System;
using System.Threading;
using AvaloniaMedia.FFmpeg.Model;
using FFmpeg.AutoGen;

namespace AvaloniaMedia.FFmpeg.Model;

public sealed unsafe class Frame : IDisposable
{
    public AVFrame* FramePointer { get; private set; } = null;
    public AVSubtitle* SubtitlePointer { get; private set; } = null;

    public bool IsNull => FramePointer == null;
    public long Pts => FramePointer->pts;
    public int Format => FramePointer->format;
    public int Width => FramePointer->width;
    public int Height => FramePointer->height;
    public int SampleRate => FramePointer->sample_rate;
    public int SampleNum => FramePointer->nb_samples;
    public int ChannelCount => FramePointer->ch_layout.nb_channels;
    public AVRational SampleAspectRatio => FramePointer->sample_aspect_ratio;
    public int BytePerSample => ffmpeg.av_get_bytes_per_sample((AVSampleFormat)FramePointer->format);
    public byte* DataPointer(uint plane) => FramePointer->data[plane];
    public int LineSize(uint plane) => FramePointer->linesize[plane];

    public double Time { get; set; } = double.NaN; // 以秒为单位
    public double Duration { get; set; } = double.NaN;
    public int Serial { get; set; } = -1;
    public bool Uploaded { get; set; } = false;
    public double Speed { get; set; } = 1.0;
    public string? Subtitle { get; set; }


    /// <summary>
    /// 创建一个初始化的Frame对象，并分配内存；初始状态下，各计算属性无效，不应访问
    /// </summary>
    /// <returns></returns>
    /// <exception cref="FFmpegException">资源创建失败异常</exception>
    public static Frame CreateFrame()
    {
        var frame = ffmpeg.av_frame_alloc();
        if (frame == null)
            throw new FFmpegException("Failed to allocate frame");

        return new Frame { FramePointer = frame };
    }

    public static Frame MoveFrame(Frame other)
    {
        var framePtr = ffmpeg.av_frame_alloc();
        if (framePtr == null)
            throw new FFmpegException("Failed to allocate frame");

        ffmpeg.av_frame_move_ref(framePtr, other.FramePointer);
        var frame = new Frame
        {
            FramePointer = framePtr,
            Time = other.Time,
            Duration = other.Duration,
            Serial = other.Serial,
            Uploaded = other.Uploaded,
            Speed = other.Speed,
            Subtitle = other.Subtitle
        };
        // other.FramePointer = null; // error: 'other' is a 'ref struct' and cannot be used as a 'variable'
        other.Time = double.NaN;
        other.Duration = double.NaN;
        other.Serial = -1;
        other.Uploaded = false;
        other.Speed = 1.0;
        other.Subtitle = null;
        return frame;
    }

    public void Unref()
    {
        ffmpeg.av_frame_unref(FramePointer);
    }

    /// <summary>
    /// 释放Frame中的资源，将其变为空状态，Frame.IsNull为true
    /// 该函数不是线程安全的
    /// </summary>
    public void Release()
    {
        if (SubtitlePointer != null)
        {
            ffmpeg.avsubtitle_free(SubtitlePointer);
            SubtitlePointer = null;
        }

        if (FramePointer != null)
        {
            var ptr = FramePointer;
            ffmpeg.av_frame_free(&ptr);
            FramePointer = null;
        }
    }

    private bool disposed = false;

    /// <summary>
    /// 回收Frame的所有资源，该函数是线程安全的
    /// </summary>
    public void Dispose()
    {
        if (Interlocked.Exchange(ref disposed, true) == true)
            return;

        Release();

        GC.SuppressFinalize(this);
    }

    ~Frame() => Dispose();
}

File: FFmpegNative/Native/Frame.cpp
#include "Frame.h"

#include <cassert>

#include "FFmpegException.h"

Frame::Frame(AVFrame *ptr)
{
    this->raw_frame_ptr = ptr;
}

Frame::Frame(Frame &&other) noexcept
{
    this->raw_frame_ptr = other.raw_frame_ptr;
    other.raw_frame_ptr = nullptr;
}

Frame &Frame::operator=(Frame &&other) noexcept
{
    assert(this != &other);
    av_frame_free(&this->raw_frame_ptr);
    this->raw_frame_ptr = other.raw_frame_ptr;
    other.raw_frame_ptr = nullptr;
    return *this;
}

Frame::~Frame()
{
    av_frame_free(&this->raw_frame_ptr);
}

Frame Frame::create()
{
    auto raw_ptr = av_frame_alloc();
    if (raw_ptr == nullptr)
    {
        throw FFmpegException("Failed to allocate frame");
    }
    auto frame = Frame();
    frame.raw_frame_ptr = raw_ptr;
    return frame;
}

Frame Frame::ref(const Frame &frame)
{
    auto result = Frame::create();
    av_frame_ref(result.get(), frame.get());
    return result;
}

AVFrame *Frame::get()
{
    return this->raw_frame_ptr;
}

const AVFrame *Frame::get() const
{
    return this->raw_frame_ptr;
}

void Frame::reset(AVFrame *ptr) noexcept
{
    if (this->raw_frame_ptr != nullptr)
    {
        av_frame_free(&this->raw_frame_ptr);
    }
    this->raw_frame_ptr = ptr;
}

bool Frame::is_null() const
{
    return this->raw_frame_ptr != nullptr;
}

int64_t Frame::pts() const
{
    assert(this->raw_frame_ptr != nullptr);
    return this->raw_frame_ptr->pts;
}

int Frame::format() const
{
    assert(this->raw_frame_ptr != nullptr);
    return this->raw_frame_ptr->format;
}

int Frame::width() const
{
    assert(this->raw_frame_ptr != nullptr);
    return this->raw_frame_ptr->width;
}

int Frame::height() const
{
    assert(this->raw_frame_ptr != nullptr);
    return this->raw_frame_ptr->height;
}

int Frame::sample_rate() const
{
    assert(this->raw_frame_ptr != nullptr);
    return this->raw_frame_ptr->sample_rate;
}

int Frame::sample_count() const
{
    assert(this->raw_frame_ptr != nullptr);
    return this->raw_frame_ptr->nb_samples;
}

int Frame::channel_count() const
{
    assert(this->raw_frame_ptr != nullptr);
    return this->raw_frame_ptr->ch_layout.nb_channels;
}

uint8_t **Frame::data()
{
    assert(this->raw_frame_ptr != nullptr);
    return this->raw_frame_ptr->data;
}

const uint8_t *Frame::data(int index) const
{
    assert(this->raw_frame_ptr != nullptr);
    return this->raw_frame_ptr->data[index];
}

uint8_t *Frame::data(int index)
{
    assert(this->raw_frame_ptr != nullptr);
    return this->raw_frame_ptr->data[index];
}

const int *Frame::linesize() const
{
    assert(this->raw_frame_ptr != nullptr);
    return this->raw_frame_ptr->linesize;
}

int *Frame::linesize()
{
    assert(this->raw_frame_ptr != nullptr);
    return this->raw_frame_ptr->linesize;
}

int Frame::linesize(int index) const
{
    assert(this->raw_frame_ptr != nullptr);
    return this->raw_frame_ptr->linesize[index];
}